rclpy/rclpy/lifecycle/node.py

```diff
@@ -0,0 +1,392 @@
+            if ret != TransitionCallbackReturn.SUCCESS:
+                return ret
+        return TransitionCallbackReturn.SUCCESS
+
+    def on_activate(self, state) -> TransitionCallbackReturn:
+        for entity in self._managed_entities:
+            ret = entity.on_activate(state)
+            if ret != TransitionCallbackReturn.SUCCESS:
+                return ret
+        return TransitionCallbackReturn.SUCCESS
+
+    def on_deactivate(self, state) -> TransitionCallbackReturn:
+        for entity in self._managed_entities:
+            ret = entity.on_deactivate(state)
+            if ret != TransitionCallbackReturn.SUCCESS:
+                return ret
+        return TransitionCallbackReturn.SUCCESS
+
+    def on_error(self, state) -> TransitionCallbackReturn:
+        for entity in self._managed_entities:
+            ret = entity.on_error(state)
+            if ret != TransitionCallbackReturn.SUCCESS:
+                return ret
+        return TransitionCallbackReturn.SUCCESS
+
+    def create_lifecycle_publisher(self, *args, **kwargs):
+        # TODO(ivanpauno): Should we override lifecycle publisher?
+        # There is an issue with python using the overridden method
+        # when creating publishers for builitin publishers (like parameters events).
+        # We could override them after init, similar to what we do to override publish()
+        # in LifecycleNode.
+        # Having both options seem fine.
+        if 'publisher_class' in kwargs:
+            raise TypeError(
+                "create_publisher() got an unexpected keyword argument 'publisher_class'")
+        pub = Node.create_publisher(self, *args, **kwargs, publisher_class=LifecyclePublisher)
+        self._managed_entities.add(pub)
+        return pub
+
+    def destroy_lifecycle_publisher(self, publisher: LifecyclePublisher):
+        try:
+            self._managed_entities.remove(publisher)
+        except KeyError:
+            pass
+        return Node.destroy_publisher(self, publisher)
+
+    def __register_callback(
+        self, state_id: int, callback: Callable[[LifecycleState], TransitionCallbackReturn]
+    ) -> bool:
+        """
+        Register a callback that will be triggered when transitioning to state_id.
+
+        The registered callback takes as an argument the previous state id, and returns
+        a TransitionCallbackReturn value.
+        """
+        self._callbacks[state_id] = callback
+        # TODO(ivanpauno): Copying rclcpp style.
+        # Maybe having a return value doesn't make sense (?).
+        # Should we error/warn if overridding an existing callback?
+        return True
+
+    def __execute_callback(
+        self, current_state_id: int, previous_state: LifecycleState
+    ) -> TransitionCallbackReturn:
+        cb = self._callbacks.get(current_state_id, None)
+        if cb is None:
+            return TransitionCallbackReturn.SUCCESS
+        try:
+            ret = cb(previous_state)
```

Comment:
Is there or is there currently a way to enforce that the user's callbacks return a return code? Not a big issue, I'm just thinking of a user forgetting to return something, so `None` is returned and then the `resp.success` member is set to `False` in `__on_change_state` even though maybe the callback ran fine.

Reply:
We can check if the return type is correct and show a nice error if it's not.
I have just made that mistake while writting the demo and the current error message is not super obvious.

Reply:
Added some extra checks in https://github.com/ros2/rclpy/pull/865/commits/bc982471c1109292b950d6c61b8f3cfbc34724a6.